Add MonadBaseControl

1Jajen1 opened the pull request:
Closes https://github.com/arrow-kt/arrow-fx/issues/27

Still missing laws, tests and docs.

When all the reorder prs are merged I can also add instances for those types.

Also this does not use kapt `extension` for two reasons: It enables the use of the default definitions and it prevents an out of memory error due to excessive generated code...

I am also trying to get a concurrent instance using this but I have one problem:
A type like `OptionT<F, A>` has a case where the `A` may not be available, this means that in `race` the race result may not be available. That is why the instances cancel all fibers if a `None` would be returned. In a fully generic setting it is impossible to know if this case happens. Apart from passing a custom callback to handle the result I have no good idea how to solve this.

nomisRev commented:
> I am also trying to get a concurrent instance using this but I have one problem:
A type like OptionT<F, A> has a case where the A may not be available, this means that in race the race result may not be available. That is why the instances cancel all fibers if a None would be returned. In a fully generic setting it is impossible to know if this case happens. Apart from passing a custom callback to handle the result I have no good idea how to solve this.

I think we need to build custom impl per data-type, similarly to how the `Concurrent` instances are implemented. I'm very interested in checking this out next week.

> Edit: Also this uses a bit of unsafe casting and wrapping only for type checking purposes. Any ideas on how to do this differently?

Need some time to dig into this!! Not sure if we can completely avoid such type casting, those are sadly common in this kind of code. Super interesting stuff! Awesome work @1Jajen1 👏 👏

1Jajen1 commented:
Still missing docs and tests, but everything else should no be there^^

1Jajen1 commented:
Tests are done. I'll do docs next

1Jajen1 commented:
Added some docs, can't really say I like them all that much, but that should do it for now.

1Jajen1 commented:
So I updated this, seems like `AtomicRefW` was removed and replaced with `Ref` on master. That sucks tho as it forces `F` to be `MonadDefer` and while it does use a mutable var it is local mutability only and hence should be considered pure.

Any ideas how this can be circumvented?

aballano commented:
> So I updated this, seems like `AtomicRefW` was removed and replaced with `Ref` on master. That sucks tho as it forces `F` to be `MonadDefer` and while it does use a mutable var it is local mutability only and hence should be considered pure.
> 
> Any ideas how this can be circumvented?

We're using our own suspended Atomic from `arrow.fx.coroutines.Atomic` which you might be able to use if I'm not mistaken cc @nomisRev

nomisRev commented:
@1Jajen1 `AtomicRefW` is a simple wrapper around `atomic-fu` which is compiler generated and erased. The reason for the wrapper is that `atomic-fu` is a compiler generator for [`AtomicReferenceFieldUpdater`](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicReferenceFieldUpdater.html).
Which means you can only use it inside a `class`.

Seems like we have two options here, encode the monad base control function as a class that inherits a function, or re-add a small internal wrapper. You can of course also always construct a `Ref` for `F` inside `F`.

We can re-consider the removal from Arrow Fx IO's module, since we're deprecating it anywhere.
We felt that promoting `atomic` over `Atomic` for classes and promoting `Atomic` for `suspend fun` would make most to promoting idiomatic FP in Kotlin.

nomisRev commented:
Also we should re-investigate this for `suspend`! Having a suspend base control would be quite powerful.

1Jajen1 commented:
> You can of course also always construct a `Ref` for `F` inside `F`

Only with access to suspend or `MonadDefer` right? And I am trying to avoid both of those ^^

Thinking about this a bit more: A normal variable should be fine as well as the only two methods that ever write to it are the use and finalizer parts. They are guaranteed to never run concurrently right? So there should never ever be a race-condition there. Or am I missing something here?

> Also we should re-investigate this for suspend! Having a suspend base control would be quite powerful.

Well that depends: `MonadBaseControl` is very useful for mtl, so as long as we have a way to do mtl on top of fx-coroutines it'll make sense. `MonadBaseControl` abstracts state handling during destruction and construction of transformer stacks, thus it is amazing when lifting methods like `bracket` and other higher order functions, but that is also quite a narrow use case (limited to transformer stacks).

But if we ever have a fast multishot version of `ContT` + some basic type level programming we'll be able to offer something much better without having the need for this typeclass (and quite frankly `MonadBaseControl` as useful as it is, it is still awful because it has some caveats for certain types of higher-order functions that it can't really lift correctly).

1Jajen1 commented:
I'll use the atomic ref from fx for now, seems to do the job just as well ^^

The future of the mtl module is quite uncertain anyway so I think this is fine for now without changes to deps etc.

1Jajen1 commented:
I don't really want to advance https://github.com/arrow-kt/arrow-check/pull/24 any further until this is either merged or https://github.com/arrow-kt/arrow/issues/2186 is resolved. The reason is that it would be a very manual task without the gradle tasks and I also don't want to install android sdks just for this.

Since it is just arrow-check failing I'd also be fine with just merging this and having it broken for the time it takes to bring the `MonadTrans` changes over (which is just the change in `MonadTransLaws` and the slight change of `MonadTrans` now having a `liftMonad` method)
